aptos-move/framework/src/natives/string_utils.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    natives::helpers::{make_safe_native, SafeNativeContext, SafeNativeError, SafeNativeResult},
+    safely_pop_arg,
+};
+use aptos_types::on_chain_config::{Features, TimedFeatures};
+use move_core_types::{
+    account_address::AccountAddress,
+    gas_algebra::{GasQuantity, InternalGas},
+    language_storage::TypeTag,
+    u256,
+    value::{MoveFieldLayout, MoveStructLayout, MoveTypeLayout},
+};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Reference, Struct, Value, Vector},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, fmt::Write, ops::Deref, sync::Arc};
+
+pub fn format_vector(
+    context: &mut SafeNativeContext,
+    base_gas: InternalGas,
+    v: Vec<(&MoveTypeLayout, Value)>,
+    out: &mut String,
+) -> SafeNativeResult<()> {
+    for (i, (ty, val)) in v.into_iter().enumerate() {
+        if i > 0 {
+            out.push_str(", ");
+        }
+        native_format_impl(context, base_gas, ty, val, out)?;
+    }
+    Ok(())
+}
+
+pub fn format_vector_with_fields(
+    context: &mut SafeNativeContext,
+    base_gas: InternalGas,
+    fields: &[MoveFieldLayout],
+    strct: Struct,
+    out: &mut String,
+) -> SafeNativeResult<()> {
+    for (i, (ty, x)) in fields.iter().zip(strct.unpack()?).enumerate() {
+        if i > 0 {
+            out.push_str(", ");
+        }
+        write!(out, "{}: ", ty.name).unwrap();
+        native_format_impl(context, base_gas, &ty.layout, x, out)?;
+    }
+    Ok(())
+}
+
+pub fn native_format_impl(
+    context: &mut SafeNativeContext,
+    base_gas: InternalGas,
+    ty: &MoveTypeLayout,
+    val: Value,
+    out: &mut String,
+) -> SafeNativeResult<()> {
+    context.charge(base_gas)?;
+    match ty {
+        MoveTypeLayout::Bool => {
```

Comment:
Yeah, the same trick which @zjma uses for the algebra :-) In the absence of traits which could resolve this switch a compile time, without need of natives, we use the native level to resolve the overloading. 

Again a good argument why we need traits in the language.